### PR TITLE
validator health check sc 

### PR DESCRIPTION
**Pull request by Hitenjain14**
## Changes
Created a sc function validator_health_check - [issue #2128 ](https://github.com/0chain/0chain/issues/2128)

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber: [validator health check worker](https://github.com/0chain/blobber/pull/951)
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/storagesc/validator.go [hunk 1]
```diff
@@ -257,3 +257,33 @@ func (sc *StorageSmartContract) updateValidator(t *transaction.Transaction,
...
+	emitValidatorHealthCheck(validator, downtime, balances)
+
+	_, err = balances.InsertTrieNode(validator.GetKey(sc.ID), validator)
+
+	if err != nil {
+		return "", common.NewError("validator_health_check_failed",
+			"can't Save validator: "+err.Error())
+	}
+
+	return (string(validator.Encode())), nil
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/validator.go]: Let's do `return string(validator.Encode()), nil`

**Review by peterlimg — CHANGES_REQUESTED** [review]
Looks good to me 👍 , please add unit tests to make sure the new added SC work as expected.

**Comment by boddumanohar** [comment]
@Hitenjain14  please see if you could also add `isActive` query parameter to /validators endpoint
https://github.com/0chain/0chain/pull/2127

**Comment by Hitenjain14** [comment]
> @Hitenjain14 please see if you could also add `isActive` query parameter to /validators endpoint #2127

Will look into it, do we have any unit tests for blobber_health_check ?

File: code/go/0chain.net/smartcontract/storagesc/validator.go (hunk 1 above)
Comment (Hitenjain14) [reply code/go/0chain.net/smartcontract/storagesc/validator.go]: Added the required changes .

File: code/go/0chain.net/smartcontract/storagesc/blobber_test.go [hunk 2]
```diff
@@ -1299,3 +1299,26 @@ func Test_flow_no_challenge_responses_cancel(t *testing.T) {
...
+
+	var (
+		blob   = addBlobber(t, ssc, 2*GB, tp, avgTerms, 50*x10, balances)
+		b, err = ssc.getBlobber(blob.id, balances)
+	)
+	require.NoError(t, err)
+
+	// check health
+	_, err = healthCheckBlobber(t, b, 0, tp, ssc, balances)
+	require.NoError(t, err)
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/blobber_test.go:1322]: check that healthcheck date is really updated

File: code/go/0chain.net/smartcontract/storagesc/validator_test.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	var (
+		validator = addValidator(t, ssc, tp, balances)
+		v, err    = getValidator(validator.id, balances)
+	)
+
+	require.NoError(t, err)
+
+	_, err = healthCheckValidator(t, v, 0, tp, ssc, balances)
+
+	require.NoError(t, err)
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/validator_test.go:29]: check that data is really updated

**Review by dabasov — APPROVED** [review]

File: code/go/0chain.net/smartcontract/storagesc/validator_test.go (hunk 3 above)
Comment (Hitenjain14) [reply code/go/0chain.net/smartcontract/storagesc/validator_test.go:29]: They are getting checked in healthCheckBlobber / healthCheckValidator as the transaction are created in these functions and we need to check with tx.CreationData

**Review by peterlimg — APPROVED** [review]
LGTM

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | fix-rounding-off-for-balance  |
| zwalletcli  | fix-rounding-off-for-balance  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | fix-rounding-off-for-balance  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | fix-rounding-off-for-balance  |
| zwalletcli  | fix-rounding-off-for-balance  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | fix-rounding-off-for-balance  |
| zwalletcli  | fix-rounding-off-for-balance  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**